src/Hosting/ResponsivePageRouteModelConvention.cs

```diff
@@ -0,0 +1,77 @@
+using System;
+using System.IO;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Wangkanai.Detection.Models;
+
+namespace Wangkanai.Detection.Hosting
+{
+    internal class ResponsivePageRouteModelConvention : IPageRouteModelConvention
+    {
+        public void Apply(PageRouteModel model)
+        {
+            var fileNameWithoutExtension = model.ViewEnginePath.Split('/')[^1];
+            if (!fileNameWithoutExtension.Contains('.'))
+            {
+                // Normal page, nothing to do!
+                return;
+            }
+
+            // This just implements the 'suffix' strategy
+            var split = fileNameWithoutExtension.Split('.');
+            if (split.Length != 2)
+            {
+                throw new InvalidOperationException($"Page '{model.RelativePath}' does not follow the required format.");
+            }
+
+            var pageName = split[0];
+            var deviceName = split[1];
+
+            if (!Enum.TryParse<Device>(deviceName, ignoreCase: true, out var device))
+            {
+                throw new InvalidOperationException($"Device name could not be parsed for page '{model.RelativePath}'.");
+            }
+
+            // Since the page name has something like `.mobile` in it, the special cased rules for Index.cshtml
+            // don't apply. We know there's going to be one selector.
+            var selector = model.Selectors.Single(); 
+
+            // Remove the device name from the route template. This is complicated because the route template
+            // can additional parameters defined in the page itself.
+            //
+            // Ex: we need to turn `About/Help.mobile/{id?}` into `About/Help/{id?}`
```

Comment:
When does this happen? Before or after routing endpoint take place.

Reply:
This is done statically while we build the route table.

Basically what we're doing here is we'll have routes like:

```text
About/Help
About/Help.tablet
About/Help.mobile
```

This code will assign all of these pages to the route `About/Help` and then it will disambiguate based on the device metadata.